### DISCUSSION TITLE
OAuth login via API

### DISCUSSION THREAD

**Discussion by zedlauzy** [apache/superset#20601] [Q&A / Help] [answered]
I have using AUTH_OAUTH with Azure and it is working well in Superset Login. I understand in FAB that we can get bearer token with user with db provider by doing the following:

```
payload = {
 'username': 'admin',
 'password': 'password',
 'provider': 'db', 
}
req = requests.post('http://localhost:8088/api/v1/security/login', json=payload)

bearer_token = {
 'Authorization': f'Bearer {req.json()["access_token"]}'
}
# Use bearer_token to access Superset APIs
```

Can we do something similar above with oauth provider? If yes, how do we get access token to access Superset APIs?

**Comment by sker65** [comment]
Same here:
I all examples logging in to superset api is done with "db" provider. Is there a way to login with an external OAuth provider as well (as the login to superset as well is handled).
I have a custom security manager in place and the superset login works well, but to access the api I still need to login via username / password using the "db" provider.
I saw that allow_browser_login is set to True for the security api, but what is needed to make it work actually? Do I need to promote a session cookie with the subsequent requests? My client in this case is not a browser but a JS service.
Is there an example for that: login in, then requesting a guest token??

**Comment by sker65** [comment, marked as answer]
What I found: it is indeed the session cookie that needs to be sent along with the request. So as soon as you are fully logged in with whatever your custom security impl. requires (OAuth, OpenID Connect) you should get a cookie called "session" along with the response. Adding this cookie in every API request is enough to grant access as long as the respective API call is marked with "allow_browser_login" and the username / roles associated with the session has all required permissions to make the call

**Reply by zedlauzy** [reply]
Thanks for sharing as this answer inspires me to create a `CustomSecurityApi.py` which has a new login API to just take in the bearer token header and go through my custom `auth_user_oauth` which returns an user and return a newly created access token with the user. I didn't require a session cookie for this approach however but thanks for your help!

**Reply by benjamin-mueller** [reply]
This is exactly what I have been trying! Do you mind sharing the code?

**Reply by giovannipapini-agilelab** [reply]
Would you mind sharing that code @zedlauzy?

**Reply by zedlauzy** [reply]
The API should be built based on your requirements: do refer to https://github.com/dpgaspar/Flask-AppBuilder/search?q=SecurityApi to build it

**Reply by ProjectsOfMLee** [reply]
Hi @sker65 , do you add the temporary cookie in the request header in each API call? or how do you add the session cookie in every API request?

**Reply by Gunaranjan66** [reply]
by  using below custom OAuth `\login` api with url token arg return **access_token** & **refresh_token**
```
# CustomSecurityManager.py
from flask import redirect, g, flash, request
from superset.security import SupersetSecurityManager
from flask_appbuilder.security.views import UserDBModelView,AuthOAuthView
from flask_appbuilder.security.views import expose
from flask_login import login_user, logout_user
from flask_jwt_extended import (
    create_access_token,
    create_refresh_token,
    get_jwt_identity,
    jwt_required,
)

class CustomOAuthView(AuthOAuthView):

    @expose('/login/', methods=['GET', 'POST'])
    @expose("/login/<provider>")
    @expose("/login/<provider>/<register>")
    def login(self, provider= None):
        token = request.args.get('token')  # update to post payload
        if token == 'SOME_TOKEN_VALUE':  # add token here
            resp = dict()
            resp['access_token'] = create_access_token(identity=5, fresh=True)
            resp['refresh_token'] = create_refresh_token(identity=5)
            return resp
        else:
            return super(CustomOAuthView,self).login(provider)



class CustomSecurityManager(SupersetSecurityManager):
    authoauthview = CustomOAuthView
    def __init__(self, appbuilder):
        super(CustomSecurityManager, self).__init__(appbuilder)
```

And in superset_config.py add

```
from custom_sso_security_manager import CustomSecurityManager
CUSTOM_SECURITY_MANAGER = CustomSecurityManager
```

**Reply by bryanjknight** [reply]
One consideration with the login code is you might be blindly trusting any JWT sent. It's worth adding additional code to verify the JWT's using JWKS and other info for the OAuth provider you're using

**Reply by Ralkion** [reply]
Further to @bryanjknight 's comment, assuming the provider follows best practices you can get the key details after a couple of calls to the provider:

```python
provider = current_app.config["OAUTH_PROVIDERS"][<provider name>]
metadata_url = provider.get("remote_app").get("server_metadata_url")
metadata = requests.get(metadata_url).json()
jwks_url = metadata.get("jwks_uri")
jwks = requests.get(jwks_url).json()
keys = jwks.get("keys")
```

The key ID is part of the header of the oauth token, so you can find the key details that corresponds to the incoming jwt and then verify that you should trust it.

Edit: I would recommend caching this so you don't look it up on every request.

**Reply by vgorea** [reply]
Hey, getting this error when calling the API with the token received from the OAuth provider in the header.
```
{
    "errors": [
        {
            "message": "invalid literal for int() with base 10: '84963db1-8bbe-4441-b707-289e0aa0fe60'",
            "error_type": "GENERIC_BACKEND_ERROR",
            "level": "error",
            "extra": {
                "issue_codes": [
                    {
                        "code": 1011,
                        "message": "Issue 1011 - Superset encountered an unexpected error."
                    }
                ]
            }
        }
    ]
}
```
